ReactWindows/ReactNative/Modules/Storage/AsyncStorageModule.cs

```diff
@@ -20,5 +39,240 @@ public override string Name
+                        {
+                            result.Add(new JArray { key, JToken.FromObject(value) });
+                        }
+                        else if (value.GetType() == typeof(string))
+                        {
+                            if ( ((string)value).StartsWith(_arr) )
+                            {
+                                result.Add(new JArray { key, JToken.Parse(((string)value).Substring(1)) });
+                            }
+                            else if (((string)value).StartsWith(_str))
+                            {
+                                result.Add(new JArray { key, ((string)value).Substring(1) });
+                            }
+                            else result.Add(new JArray { key, JValue.CreateNull() });
+                        }
+                    }
+                }
+            }
+            callback.Invoke(result);
+        }
+
+        /// <summary>
+        /// Inserts multiple (key, value) pairs. The insertion will replace conflicting (key, value) pairs.
+        /// </summary>
+        /// <param name="keyValueArray">Array of key values</param>
+        /// <param name="callback">Callback.</param>
+        [ReactMethod]
+        public void multiSet(JArray keyValueArray, ICallback callback)
+        {
+            if (keyValueArray == null)
+                throw new ArgumentNullException(nameof(keyValueArray));
+
+            var result = new JArray();
+
+            foreach (var keyValue in keyValueArray)
+            {
+                if (keyValue.Type == JTokenType.Array && keyValue.ToObject<JArray>().Count == 2)
+                {
+                    var pair = keyValue.ToObject<JArray>();
+                    if (pair.First.Type == JTokenType.String)
+                    {
+                        var key = pair.First.ToObject<string>();
+                        switch (pair.Last.Type)
+                        {
+                            case JTokenType.Null:
+                                _container.Values[key] = _nul;
+                                break;
+                            case JTokenType.Boolean:
+                                _container.Values[key] = pair.Last.ToObject<bool>();
+                                break;
+                            case JTokenType.Integer:
+                                _container.Values[key] = pair.Last.ToObject<long>();
+                                break;
+                            case JTokenType.Float:
+                                _container.Values[key] = pair.Last.ToObject<double>();
+                                break;
+                            case JTokenType.String:
+                                _container.Values[key] = _str + pair.Last.ToObject<string>();
+                                break;
+                            case JTokenType.Array:
+                                _container.Values[key] = _arr + pair.Last.ToString();
```

Comment:
Doesn't necessarily have to be the fastest.  Cheapest can also mean least GC impact, least total heap size impact, etc.  Now I'm equally concerned that the new approach doubles the amount of space taken up by property keys.

---
In reply to: [55735346](https://github.com/CatalystCode/react-native/pull/247#discussion_r55735346) [](ancestors = 55735346)